test/runnable/previewin.d

```diff
@@ -0,0 +1,188 @@
+    WithDtor     a10 = WithDtor(&isTestOver);
+
+    testin1(a1);
+    testin2(a2);
+    testin3(a3);
+    testin3(a5);
+    testin4(a4);
+    testin4(a6);
+    testin5(a7);
+    testin6(a8);
+    testin7(a9);
+    isTestOver = false;
+    testin8(a10, null);
+
+    // Arguments are all values, no `ref` needed
+    testin9(int.init);
+    testin9(char.init, ubyte.init, short.init, int.init, size_t.init);
+    // Arguments are all refs
+    testin9(a2, a4, a5, a6, a7, a8, a9, a10);
+    testin9(NonCopyable(true), WithPostblit(true), WithCopyCtor(true));
+    // Mixed values and ref
+    testin9(char.init, ubyte.init, a2, a4, a5, a6, a7, a8, a9, a10, size_t.init);
+
+    // With dtor
+    isTestOver = false;
+    testin10(&isTestOver, NonCopyable(true), WithPostblit(true),
+             WithCopyCtor(true), WithDtor(&isTestOver));
+    isTestOver = true;
+}
+
+void testForeach() @safe pure
+{
+    int testCallNC (in NonCopyable k, in NonCopyable v)
+    {
+        assert(k == v);
+        return k.value - v.value;
+    }
+
+    NonCopyable[NonCopyable] nc;
+    nc[NonCopyable(0)] = NonCopyable(0);
+    nc[NonCopyable(42)] = NonCopyable(42);
+    nc[NonCopyable(int.min)] = NonCopyable(int.min);
+    nc[NonCopyable(int.max)] = NonCopyable(int.max);
+    foreach (ref k, const ref v; nc)
+    {
+        assert(k.value == v.value);
+        assert(testCallNC(k, v) == 0);
+    }
+    assert(nc == nc);
+    assert(nc.length == 4);
+
+    RefT[RefT] rt;
+    rt[RefT(42)] = RefT(42);
+    rt[RefT(4)] = RefT(4);
+    rt[RefT(242)] = RefT(242);
+    rt[RefT(24)] = RefT(24);
+    foreach (k, v; rt)
+        assert(k.value == v.value);
+    assert(rt == rt);
+
+    static struct Msg
+    {
+        ubyte[3] value;
+        const(char)[] msg;
+    }
+
+    static void testMsg (in Msg k_func, in Msg v_func)
+    {
+        assert(k_func.value == v_func.value);
+        assert(k_func.msg == v_func.msg);
+        assert(k_func == v_func);
+    }
+
+    Msg[Msg] msg;
+    msg[Msg([1, 2, 3], "123")] = Msg([1, 2, 3], "123");
+    msg[Msg([42, 4, 2], "4242")] = Msg([42, 4, 2], "4242");
+    msg[Msg([242, 4, 0], "2424")] = Msg([242, 4, 0], "2424");
+    foreach (ref k_loop, ref v_loop; msg)
+        testMsg(k_loop, v_loop);
+}
+
+struct ValueT { int value; }
+struct RefT   { ubyte[64] value; }
+
+struct NonCopyable
+{
+    @safe pure nothrow @nogc:
+
+    int value;
+    this(int b) { this.value = b; }
+
+    @disable this(this);
+    @disable this(ref NonCopyable);
+}
+
+struct WithPostblit
+{
+    int value;
+    this(this) @safe pure nothrow @nogc { assert(0); }
+}
+
+struct WithCopyCtor
+{
+    @safe pure nothrow @nogc:
+
+    int value;
+    this(int b) { this.value = b; }
+    this(ref WithCopyCtor) { assert(0); }
+}
+
+struct WithDtor
+{
+    bool* value;
+    ~this() scope @safe pure nothrow @nogc {  assert(*value); }
+}
+
+@safe pure nothrow @nogc:
+
+// By value
+void testin1(in uint p) {}
+// By ref because of size
+void testin2(in ulong[2] p) {}
```

Comment:
FYI, SPARC 64-bit would pass it by value. (you need a type > 32 bytes :-)

Reply:
Good catch